### ISSUE TITLE
Concourse should support imposing limits on container resources

### ISSUE THREAD

**Issue by dsharp-pivotal** [concourse/concourse#787]
We've been having some issues lately with Concourse worker instances becoming unresponsive. This morning, we saw one and looked into the console logs to see why. It showed an "oom kill" of postgres, which is frequently run in our Concourse tasks. After talking to Garden team, thinking that a container that ooms should not cause system instability, we realized that Concourse is not asking Garden to impose a memory limit. Adding this limit (and probably others) would help with system stability when tasks are not necessarily well-behaved.

@schubert @cjcjameson 

#### Symptoms

* Concourse shows a broken connection to a worker in a resource / check step.
* `bosh vms --vitals` shows "unresponsive agent"
* AWS instance status checks shows: "Instance reachability check failed at November 19, 2016 at 1:10:00 AM UTC-8 (2 days, 10 hours and 59 minutes ago)" 

After rebooting, AWS console shows:
```
[112981.236574] Out of memory: Kill process 3295 (postgres) score 7 or sacrifice child
[112981.240097] Killed process 6174 (postgres) total-vm:528776kB, anon-rss:8552kB, file-rss:15160kB
[113160.462177] INFO: task monit:2203 blocked for more than 120 seconds.
[113160.465787]       Not tainted 4.4.0-47-generic #68~14.04.1-Ubuntu
[113160.469040] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
```
Followed by several more similar "blocked for more than 120 seconds` messages.

* Concourse version: 2.5.0
* Deployment type (BOSH/Docker/binary): BOSH

**Comment by jtarchie** [comment]
Imposing a limit of memory causing a new dimension with scheduling.

* If a container is suppose to run and no worker can meet the container requirement.
* If a worker can make the requirement, but it doesn't have the cached resources.
* If a container requests too much memory no worker can achieve.
* etc...

Concourse get's around these issues by just not worrying about it.

I'm all for a flexible scheduling system, but it is a hard problem. 

These are more notes to myself to revisit.

**Comment by dsharp-pivotal** [comment]
It's a hard problem, but it's a problem that's been solved by the likes of Kubernetes. Concourse could run containers on Kubernetes or a similar container orchestrator. That would be a pretty huge change for Concourse, but might be less work than rolling your own scheduler.

**Comment by tushar-dadlani** [comment]
To @dsharp-pivotal's point, even diego does container scheduling. It depends on what philosophy is concourse aligned to, is it meant for only maintaining state of various components of the build systems or is it also meant to do actual things on it, eg: compilation.

**Comment by jtarchie** [comment]
Sorry, if I mislead here. I am not on the concourse team. :-P

**Comment by andrewedstrom** [comment]
See the discussion at #954 for more. I closed that because it was basically the same issue.

#1284 is blocked on this

**Comment by JohannesRudolph** [comment]
Just a crazy idea, but I've been playing with [Cloud Foundry tasks](https://docs.cloudfoundry.org/devguide/using-tasks.html) lately. They are schedule by Diego, so that would make them a possible "container backend" for Concourse as well. We have access to CF and K8s Clusters, both would make excellent choices to plug into Concourse as a scheduler for running and distributing containers.

**Comment by jtarchie** [comment]
@JohannesRudolph, in concept they work, but in practice the limitations imposing are not worth it. For example, these are meant for the droplets created for apps, which have a size limit of 2GB on most distributions of CF. It has to be created before any ask can be run. There's no opportunity to mount other volumes into it afterwards.

**Comment by topherbullock** [comment]
Let's keep this issue focused on the general case of initial problem outlined: Concourse doesn't ask Garden (or whatever container engine the workers use) to impose **any** limits on containers. I've updated the issue name to reflect this general case.

Placement, management, orchestration of containers, and the technology we use is an entirely different conversation; one largely captured in #675, which discusses 'Better distribution of jobs across workers', lets not conflate the two issues here.

---

Focusing on resource limits on the containers: I believe we can come up with a solution to support imposing limits on containers in Concourse which is agnostic of the backend we use. In a previous comment on #954 (https://github.com/concourse/concourse/issues/954#issuecomment-306892747) I brought forth the idea of allowing tasks to configure `garden_resource_limits:` but I think this should be a more general.

We could mirror the [properties Garden allows for container limits](https://github.com/cloudfoundry/garden/blob/master/client.go#L176-L182), but be wary of keeping everything agnostic of the actual containerization engine underneath (since there are many backend implementations for Garden, I'm sure those settings in Garden are a "supply these and hope the backend supports it" deal. ) . Addressing Kubernetes and Diego; yes, both support limits, but how one sets those limits in a task or pipeline is still a feature we need to add; the effort here is not on implementing this in Garden or interfacing with Garden at all. 

Since task containers are the ones users have the most control over, I think we should allow task configs to contain a `limits:` section like so:
```yaml
platform: linux
limits:
  cpu:
    shares: 5
  memory:
    bytes: 1024
  # not a complete list, but you get the idea here

```

**Comment by topherbullock** [comment]
There's a few issues mentioning this usage, so I'm pinging folks so they can bring up any specific limits that are of burning importance.}

Other community members with supporting cases :
@kmacoskey - https://github.com/concourse/concourse/issues/954#issuecomment-306848524
@george - https://github.com/concourse/concourse/issues/954#issue-212531919
@vito - one of the points in https://github.com/concourse/concourse/issues/366#issue-147466782
@aahlenst https://github.com/concourse/concourse/issues/1284#issue-235496627

**Comment by EugenMayer** [comment]
fighting myself with running concourse workes on vultr on rancheros the worker basically kills itself running oom right away on a 2GB instance in no time.

I have seen a couple of suggestions, most of them ending with limiting using the max-containers setting - but that is hardly a sane way. The problem with max-containers are all the check-containers, which count as a full container, but are very resource lightweight - but there are many. And one has no control how many of those and how many job containers ( task containers ) are created.

Right now, i am basically now able to run concourse on such an instance at all, since i am affectively not able to limit the worker resources right now. 

I added those 2 here, https://github.com/EugenMayer/docker-rancher-extra-catalogs/blob/master/templates/concourseci-cluster/4/docker-compose.yml.tpl#L153

if guess `CONCOURSE_GARDEN_CPU_QUOTA_PER_SHARE` is dockers `--cpu-quota` or `--cpu-shares`. The docs say:

`Maximum number of microseconds each cpu share assigned to a container allows per quota period (default: 0) [$CONCOURSE_GARDEN_CPU_QUOTA_PER_SHARE]`

So i suggest  `--cpu-quota`. 

There is no real way to limit memory usage AFAICs. I found
`Set hard limit for the tcp buf memory, value in bytes (default: 0) [$CONCOURSE_GARDEN_TCP_MEMORY_LIMIT]`

but thats just the TCP buffer, not the overall memory limit.

I think, beside all that hard limits which should then kill containers inside the worker and make builds failed, we should be able to define ( separately ).

 a) how many check-resources are allowed
 b) how many builds at a time are allowed

While b) is the most important one in terms of resource usage

**Comment by JohannesRudolph** [comment]
Two quick thoughts on @EugenMayer's comment:

**Comment by ghostsquad** [comment]
From the sound of it, that means concourse has no concept of a "build queue" either? Which means it would be very easy to overload workers with too many jobs?

**Comment by cjcjameson** [comment]
@ghostsquad if I understand correctly, yes: Concourse doesn't really think of it as a build queue. And yes, it's easy to overload the system overall.

To go into more detail: there's a concept of "Max in flight" that you can configure for any job, which can throttle that job from overloading things. You can go so far as to set that on every job separately, but in practical terms, setting it on the natural throttle points in your pipeline works fine.

"Max in flight", in turn, is one of the conditions that can cause a build of a job to be "pending". The concept of "pending" is kinda like a build queue, perhaps.

But overall, the focus is more on the pipeline rather than a build queue. I'm guessing that Concourse deliberately defaults to not throttling / queueing because it's wanting to default for the continuous part of ideal "CI".

Regarding overloading: in an ideal situation, Concourse tries to load-balance various new builds onto all workers. But it also takes into account trying to reuse cached resources on a particular worker. So indeed, it's very easy to overload the whole system, but overloading a particular worker is subject to more details around the pipeline and Concourse deployment.

To some people, build queues might be something to which you explicitly submit a job (maybe?). Because resources want order-able versions of the same essential input, Concourse doesn't have a natural way to handle arbitrary build requests. Put another way, you never give Concourse anything: a pipeline is configured to look for specific things in specific places, and take action if there are new things there. See further discussion around the "Pull Request Resource", as it's a good illustrative example.

**Comment by ddadlani** [comment]
We are currently working to add a setting in the task config to set memory and CPU limits in the garden container. We are running into an issue where the user needs to know a lot about worker configuration to use this setting. Specifically, these are the considerations:

1) The user needs to know the minimum and maximum memory limits possible for each worker.
2) The operator needs to know whether the host running garden has swap enabled, and they need to make a decision regarding whether or not to [disable it](https://github.com/cloudfoundry/guardian/blob/ff02df0fe20a202daeae124fd44c74da847b9a11/guardiancmd/command.go#L275).

The current configuration of the `concourse-docker-image` does not allow us to set this parameter, so we need to surface that to the operator to allow memory limits to be set.

**Comment by JohannesRudolph** [comment]
So in the current iteration the workflow for a pipeline author/concourse operator is like this?
- operator configures workers with maximum memory allocatable to task containers (e.g. 16 GiB machine RAM - 1 GiB for OS). This is an operator choice. 
- pipeline author specifies task RAM and CPU limit in task.yml

Do you have any plans to make use of this info for scheduling? Even a very simple scheduler could prove very useful. I'm still eager to see something like a [K8s worker engine](https://github.com/concourse/concourse/issues/787#issuecomment-287397256), but a very simple scheduler that uses task limits to decide where to put a container could be a good 80/20 solution. We currently need to use [random worker scheduling](https://github.com/concourse/concourse/issues/1741) to work around worker gravity killing our cluster. 

Additional ponts to consider
- should resources containers also have CPU/RAM limits?
- should containers without explicitly specified limits receive a "default" (e.g. 1 GiB like in CF)? 

Concourse is moving to 4.0, so I think that would be an opportunity to bring a "breaking" change like this.

**Comment by ddadlani** [comment]
Hi @JohannesRudolph, at this time we only plan to include the option to add these limits. We are not planning to make any major scheduling changes.

We invite you to open another issue for the scheduling functionality you mention and we can certainly discuss your use case there.

**Comment by ddadlani** [comment]
This issue covers setting CPU and memory limits for garden containers per task. There is another related issue, #1284, which covers operator ability to set these limits per team.

**Comment by topherbullock** [comment]
It works!!

Created a task config with the following `container_limits`:
```
      container_limits:
        cpu: 256
        memory: 200mb
```

and hijacking the container shows:

```
root@a6ee1067-d729-4f5c-5d4a-9a95ebf743b3:/# cat /sys/fs/cgroup/cpu/cpu.shares
256
root@a6ee1067-d729-4f5c-5d4a-9a95ebf743b3:/# cat /sys/fs/cgroup/memory/memory.limit_in_bytes
209715200
root@a6ee1067-d729-4f5c-5d4a-9a95ebf743b3:/# cat /sys/fs/cgroup/memory/memory.memsw.limit_in_bytes
209715200
```

**Comment by marco-m** [comment]
Hello @topherbullock, since this issue is closed, is the documentation for `container_limits:` covered by another ticket ? :-)

**Comment by vito** [comment]
@marco-m good point - `limits` feels more general

**Comment by vito** [comment]
@marco-m fyi we're intentionally leaving this undocumented for a bit while we work out longer-term plans. but it's there and should work if you want to try it out.

**Comment by marco-m** [comment]
Thanks @vito for the details, and tanks @ddadlani for the implementation :-)
